.travis.yml

```diff
@@ -146,17 +146,13 @@ jobs:
   - <<: *stage_deploy
     if: tag IS present
     <<: *python_3_7_mixture
-    install: skip
-    script: skip
-    deploy:
-      provider: pypi
-      on:
-        all_branches: true
-      user: jaraco
-      password:
-        secure: RAfz06AINvz7bfij/YhfkAreRqamgxS8a6jSRNxntYhtJke3ZszUbIDag8+n1I+G5XT2LnMhHqPNR7Plc+AeMz7VXTuy+b81Li5kse20NYlPhd7mBVmTUpXtqYQashV5J39F4qkATBLznCOrMEomM07VTXjO/o2hmQuXniab2Uo=
-      distributions: dists
-      skip_cleanup: true
+    before_script: skip
+    env:
+    - TWINE_USERNAME=jaraco
+    # TWINE_PASSWORD
+    - secure: ahqqwT5dPGZMyOr2NyavsSQAaWgE/RKy0bD0bT55MdXfVh7lpXU6sVupxKJT6/uwS8uaAahvxBjr6mTvTs7pHuUPJ0MVsFb+LnQVUzlqc9Z47tE7nWGn/mYXb+mxN/VBhlUX/Y4rQMxcW6VFqOMSQBPQ2EukIh76nerbvQC1ypA=
+    - TOX_TESTENV_PASSENV="TWINE_USERNAME TWINE_PASSWORD"
+    script: tox -e release
```

Comment:
Also, script step corresponds to testing. It's logical to put any validation here, but not publishing.

Reply:
I disagree. Although the original use-case for tox is to invoke tests, it's actually more useful when thought of as an environment builder and command invoker, not dissimilar to what the CI tools do. By moving the release logic to tox, it makes it possible to cut releases locally using the exact same command as the CI environment does (if as happens from time-to-time the CI environment fails to deploy).

Reply:
I'm not talking about tox but Travis' `script` step vs `deploy` step which can also be described in tox of course.

As for CI deployment failures, if tests fail you can manually trigger deployment job only. Also, if it fails because of name collision (because of reruns?) there's a param to ignore such failures.